Classes/common/HTTP/CDTURLSession.m

```diff
@@ -53,8 +56,28 @@ - (NSURLSessionDataTask *)dataTaskWithRequest:(NSURLRequest *)request
                                                         NSURLResponse *response,
                                                         NSError *error))completionHandler
 {
+    
+    //make the context
+    CDTURLSessionInterceptorContext *ctx = [[CDTURLSessionInterceptorContext alloc ]initWithRequest:[request mutableCopy]];
+    for ( id obj in self.requestInterceptors) {
+        
+        if( [obj conformsToProtocol:@protocol(CDTURLSessionInterceptor)]){
+            id<CDTURLSessionInterceptor> interceptor = obj;
+            
+            if([interceptor respondsToSelector:@selector(interceptRequestInContext:)]){
+                ctx = [interceptor interceptRequestInContext:ctx];
+            } else {
+                CDTLogWarn(CDTREPLICATION_LOG_CONTEXT, @"%@ doesnn't respond to \
+                           interceptRequestInContext: skipping",interceptor);
+            }
+        } else {
+            CDTLogWarn(CDTREPLICATION_LOG_CONTEXT, @"%@ doesn't conform to protocol \
+                       CDTURLSessionInterceptor, skipping",obj);
+        }
+    }
+    
     __weak CDTURLSession *weakSelf = self;
-    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:request
+    NSURLSessionDataTask *task = [self.session dataTaskWithRequest:[ctx.request copy]
```

Comment:
I'd read that differently -- an identical array contains the same objects as are in the array. If the contained objects are mutable, they can be changed from anywhere that has access to that array. Another piece of supporting evidence, is what could be done with objects that are not `NSCopying` compliant?